### PR TITLE
refactor: Inherit from commands.Bot to achieve shared state behaviour

### PR DESCRIPTION
**Pull request by lukadd16**
This PR addresses #10

This PR adds a more elegant way of achieving the same behaviour as the current SharedState implementation, by creating a sub-class in main.py that inherits from commmands.Bot and adds some additional instance variables (which correlate to the attributes the SharedState class contained).
* Any instance variables that are set in the PCParadiseBot class can be accessed within the same class via `self.var_name` AND can be accessed in other cogs via `self.bot.var_name` (you can see being done in `help.py`).
* This is possible due to the nature of inheriting `commands.Bot` and the fact that each cog's constructor takes in `bot` as an instance variable.

I've largely left the config setup logic untouched and simply relocated the methods from `shared_state.py` and added them as static methods of our child class in `main.py`.

The other methods that I've added to the child class are fairly self-explanatory, though I will point out that I am using the proper [`load_extension()`](https://discordpy.readthedocs.io/en/stable/ext/commands/api.html#discord.ext.commands.Bot.load_extension) method from the library to load all our cogs rather than the current `importlib` implementation where you were loading in the cogs as python modules (which I'm sure has a few negative side-effects).

Last item, you may notice that I'm explicitly defining what [intents](https://discordpy.readthedocs.io/en/stable/intents.html) our bot needs. For simplicity's sake (plus the fact that you'll most likely need them when more features are added), I've subscribed to all of the intents.
* **Note:** when testing the bot and running it in a production environment, make sure you have both of the privileged intents enabled in the [Discord Developer Portal](https://discord.com/developers/applications/). If you haven't done this before, follow this [guide](https://discordpy.readthedocs.io/en/stable/intents.html#privileged-intents) from the Discord.py documentation.

### CODE REVIEW

**Comment by awsomearvinder** [comment]
Appreciated! I like that you used the static typing. :) I'll take a look at the code and review it tomorrow probably.

**Comment by Ganoosh** [comment]
I really should have dived a bit more into the library before pushing #10, although with that said, I checked over the code. I'll give vinder one last chance to review it, but looks good in my opinion. Thanks for the help

**Comment by lukadd16** [comment]
Yeah, no problem :) I've been using discord.py for over half a year now and have developed a good understanding of _almost_ everything the library has to offer, so I'd be happy to help out with the specifics for whatever you are attempting to implement.

**Review by awsomearvinder — CHANGES_REQUESTED** [review]

File: src/main.py [hunk 1]
```diff
@@ -1,51 +1,165 @@
...
+        Overrides DPY's event loop initialization logic allowing for more fine control.
+        """
+        try:
+            self.loop.run_until_complete(self.start(self.config.get("token")))
+        except KeyboardInterrupt:
+            print("\nKeyboard Interrupt Detected")
+            self.loop.run_until_complete(self.logout())
+        finally:
+            self.loop.run_until_complete(self.logout())
+            print("\nConnection Closed")
```
[the remark this answers is not in the document]
Comment (lukadd16) [reply src/main.py:133]: Good catch. Now that I look at it again I don't even know why I put in that statement twice.

File: src/main.py [hunk 2]
```diff
@@ -1,51 +1,165 @@
...
+        Blocking method that facilitates loading of all the
+        bot's cogs along with any other preliminary setup.
+        """
+        print("Loading cogs...")
+        for extension in EXTENSIONS:
+            try:
+                self.load_extension(extension)
+                print(f"SUCCESS - {extension}")
+            except commands.ExtensionNotFound:
+                print(f"FAILED - {extension}")
```
[the remark this answers is not in the document]
Comment (lukadd16) [reply src/main.py:120]: Got it

File: src/main.py [hunk 3]
```diff
@@ -1,51 +1,165 @@
...
+    Sub-class that inherits from commands.Bot to add additional attributes
+    and have finer control over certain aspects of the bot.
     """
-    try:
-        config = SharedState("config.ini")
-    except FileNotFoundError:
+
+    def __init__(self):
+        self.config = PCParadiseBot.initialize_config()
+        self.launch_time = datetime.utcnow()
```
[the remark this answers is not in the document]
Comment (lukadd16) [reply src/main.py:32]: I explained this in the description of my first commit. Currently the `launch_time` attribute isn't used anywhere else in the codebase, but I thought at some point we would need it to calculate the bot's uptime (i.e. for use in a command).

**Comment by awsomearvinder** [comment]
I'll merge this in, I tried resolving the merge conflict directly from the github webpage because I didn't have my development enviornment with me - looks like black and pylint didn't like it.

**Comment by lukadd16** [comment]
I've resolved the issues in my latest commit

**Pull request merged**